File: nodejs/systemtest/impersonation.test.ts
/**
 * Sessions acting as an external user, against the real service.
 *
 * An API client places its own users in customer networks, then opens memory
 * sessions as them. The fake server in `tests/` already proves which key each
 * call carries; what only the real service can prove is what those keys do. A
 * write made through a session is the user's: the user finds it, and the user
 * alone can revert it, since a revert answers only for its caller's own
 * writes. Two sessions running at once each act as their own user, each in a
 * network of its own, so a crossed key would show as one finding what the
 * other wrote. And closing a session really revokes its key.
 *
 * The memories written follow the lifecycle suite's rules, for the same
 * reasons: the run marker lives only in the query, which becomes the memory's
 * intent, and the insight is plain prose about this SDK, which the quality gate
 * accepts. Each subject is distinct, and about the Node.js SDK alone, so no
 * write reads as the same knowledge as another — this one's or another SDK's.
 * A write a failing test leaves behind goes with its customer network, whose
 * deletion takes the memories placed in it.
 */

import assert from 'node:assert/strict'
import { test } from 'node:test'
import { setTimeout as sleep } from 'node:timers/promises'

import {
  DataSource,
  Memco,
  MemcoAuthenticationError,
  MemcoNotFoundError,
  RevertOutcome
} from '../src/index.js'
import type { Memory } from '../src/index.js'
import type { Minted } from '../src/internal/auth.js'
import * as pb from '../src/internal/gen.js'
import type { Caller } from '../src/operations.js'
import {
  Created,
  INGEST_TIMEOUT_MS,
  NO_CLIENT,
  adminClient,
  hasClient,
  pollWaits,
  rootNetwork,
  withAdmin
} from './support.js'

interface Fields {
  query: string
  title: string
  content: string
}

const OPENING: Fields = {
  query:
    "How does the Memco Node.js SDK open a memory session as one of an organisation's own users?",
  title: 'Node.js SDK impersonated sessions',
  content:
    "The Memco Node.js SDK opens a memory session as one of an organisation's own users " +
    'when `startSession` or `withSession` is given an `externalId`. It mints an ' +
    "impersonation key under the API client's token, lists the domains under that key " +
    'so the session learns its limits as that user, and only then starts the session.\n\n' +
    "Every call made through the session carries that key rather than the client's own " +
    'token. Closing the session, or leaving its `await using` block, ends the key on the ' +
    'service instead of leaving it live until it expires.'
}
const RENEWAL: Fields = {
  query:
    'When does the Memco Node.js SDK renew the impersonation key a session holds?',
  title: 'Node.js SDK impersonation key renewal',
  content:
    'The Memco Node.js SDK renews an impersonation key once four fifths of its lifetime ' +
    'have passed. Calls arriving together at that point share a single mint, and share its ' +
    'failure too, and the key it replaces is ended only after the last call still carrying ' +
    'it has finished, so a renewal never revokes a key under a call in flight.'
}
const SWEEP: Fields = {
  query:
    'What happens to open impersonation keys when a Memco Node.js SDK client closes?',
  title: 'Node.js SDK impersonation key cleanup',
  content:
    'Closing a Memco Node.js SDK client ends every impersonation key a session left open. ' +
    "It waits for calls in flight to finish, then ends each key under the client's own " +
    'token, counting a key the service reports as already gone as ended. A session dropped ' +
    'without closing has its key ended once it has been garbage-collected, by the next call ' +
    'the client makes.'
}

/** The fields to write, with the run marker added to the query alone. */
function asked(fields: Fields, marker: string): Fields {
  return { ...fields, query: `${fields.query} (system test ${marker})` }
}

/**
 * Whether a memory was written with this marker in its query.
 *
 * The query passed to createMemory becomes one of the memory's intents.
 */
function carries(memory: Memory, marker: string): boolean {
  return memory.intents.some(intent => intent.includes(marker))
}

/**
 * Collect every impersonation key the client mints from here on.
 *
 * Private access, deliberately: the SDK exposes a key's value nowhere, which
 * is the point of it, and only the value lets the test ask the service itself
 * whether the key still works once its session is closed. Each is kept as the
 * SDK holds it, whose renderings leave the value out, so an assertion printing
 * one on failure cannot put a live key in the CI log.
 */
function recordKeys(client: Memco): Minted[] {
  const keys: Minted[] = []
  const internals = client as unknown as {
    mintKey(externalId: string, owner: object): Promise<Minted>
  }
  const mint = internals.mintKey.bind(client)
  internals.mintKey = async (externalId, owner) => {
    const key = await mint(externalId, owner)
    keys.push(key)
    return key
  }
  return keys
}

/** Whether the service still accepts a key, as a client authenticating with it. */
async function accepted(key: Minted): Promise<boolean> {
  const probe = new Memco({ token: key.value })
  try {
    await probe.connect()
    return true
  } catch (error) {
    if (error instanceof MemcoAuthenticationError) {
      return false
    }
    throw error
  } finally {
    await probe.close()
  }
}

/**
 * Poll until the user finds the memory carrying the marker, failing the moment
 * they find what `other` wrote.
 *
 * A fresh session per attempt: within one session a memory already delivered
 * comes back as a bare reference carrying no intents, so a second search in
 * the same session could never recognise it.
 */
async function foundAlone(
  client: Memco,
  domain: string,
  user: string,
  query: string,
  other: string | null
): Promise<void> {
  const deadline = Date.now() + INGEST_TIMEOUT_MS
  const waits = pollWaits()
  while (Date.now() < deadline) {
    await using session = await client.memory.withSession(domain, {
      externalId: user
    })
    const { memories } = await session.search(query)
    if (other !== null) {
      assert.ok(
        !memories.some(memory => carries(memory, other)),
        `${user}'s session found what ${other} wrote, in a network ${user} is not in`
      )
    }
    if (memories.some(memory => carries(memory, user))) {
      return
    }
    console.log(`  waiting for ingestion of ${user}'s memory`)
    await sleep(waits.next().value)
  }
  assert.fail(
    `${user}'s memory never became searchable within ${INGEST_TIMEOUT_MS / 1000}s`
  )
}

/** Write, find and revert a memory as one user, never seeing the other's. */
async function act(
  client: Memco,
  domain: string,
  user: string,
  fields: Fields,
  other: string | null
): Promise<void> {
  await using session = await client.memory.withSession(domain, {
    externalId: user
  })
  const before = (await session.search(fields.query)).memories
  assert.ok(!before.some(memory => carries(memory, user)))
  const write = await session.createMemory({
    query: fields.query,
    title: fields.title,
    content: fields.content,
    source: DataSource.AGENT
  })
  assert.ok(
    write.operationId,
    'the write was accepted without an operation id to revert'
  )
  await foundAlone(client, domain, user, fields.query, other)
  const undo = await session.revertMemory(write.operationId)
  assert.equal(
    undo.outcome,
    RevertOutcome.MEMORY_REMOVED,
    `reverting as ${user}, who wrote it, reported ${RevertOutcome[undo.outcome]}`
  )
}

const skip = hasClient() ? false : NO_CLIENT

test(
  'a session acts as an external user, and its key ends with it',
  { skip },
  async () => {
    await withAdmin(async admin => {
      const created = new Created(admin, await rootNetwork(admin))
      try {
        const [network, user] = await created.placed()
        const keys = recordKeys(admin)
        console.log(`\n[${network.domain}] acting as ${user.externalId}`)

        const fields = asked(OPENING, user.externalId)
        {
          await using session = await admin.memory.withSession(network.domain, {
            externalId: user.externalId
          })
          const before = (await session.search(fields.query)).memories
          assert.ok(!before.some(memory => carries(memory, user.externalId)))
          const write = await session.createMemory({
            ...fields,
            source: DataSource.AGENT
          })
          assert.ok(
            write.operationId,
            'the write has no operation id to revert'
          )
          await foundAlone(
            admin,
            network.domain,
            user.externalId,
            fields.query,
            null
          )
          const undo = await session.revertMemory(write.operationId)
          assert.equal(
            undo.outcome,
            RevertOutcome.MEMORY_REMOVED,
            `reverting as the user who wrote it reported ${RevertOutcome[undo.outcome]}`
          )
          // The key works while its session is open, so the refusal below is
          // the close's doing rather than a key that never worked from outside.
          assert.ok(await accepted(keys[0] as Minted))
        }

        const [first] = keys
        assert.ok(first !== undefined)
        assert.ok(
          !(await accepted(first)),
          'the key still works after its session closed'
        )
        // Ending an ended key is answered NOT_FOUND, which the SDK counts as
        // ended already.
        const internals = admin as unknown as {
          call: Caller
          transport: { admin: pb.admin.AdminServiceClient }
        }
        const stub = internals.transport.admin
        await assert.rejects(
          internals.call(
            stub.endImpersonation.bind(stub),
            pb.admin.EndImpersonationRequest.fromPartial({
              externalId: user.externalId,
              keyId: first.keyId
            }),
            undefined,
            'EndImpersonation'
          ),
          MemcoNotFoundError
        )
      } finally {
        await created.remove()
      }
    })
  }
)

test(
  'two users act at once, each under a key of its own',
  { skip },
  async () => {
    await withAdmin(async admin => {
      const created = new Created(admin, await rootNetwork(admin))
      try {
        const [network, first] = await created.placed()
        const [, second] = await created.placed()
        const [one, two] = [first.externalId, second.externalId]
        console.log(`\n[${network.domain}] acting as ${one} and ${two} at once`)

        const client = adminClient()
        try {
          await client.connect()
          const keys = recordKeys(client)
          await Promise.all([
            act(client, network.domain, one, asked(RENEWAL, one), two),
            act(client, network.domain, two, asked(SWEEP, two), one)
          ])
          // Each checked before the client closes, so what ended it is its own
          // session's close rather than the client's sweep.
          assert.ok(keys.length >= 2, `${keys.length} keys were minted`)
          for (const key of keys) {
            assert.ok(!(await accepted(key)), `key ${key.keyId} still works`)
          }
        } finally {
          await client.close()
        }
      } finally {
        await created.remove()
      }
    })
  }
)

File: go/systemtest/impersonation_test.go
//go:build systemtest

package systemtest

// Sessions acting as an external user, against the live service. The fake
// server already proves which key each call carries; what only the real
// service can prove is what those keys do. A write made through a session is
// the user's: the user finds it, and the user alone can revert it, since a
// revert answers only for its caller's own writes. Two sessions running at
// once each act as their own user, each in a network of its own, so a crossed
// key would show as one finding what the other wrote. And closing a session
// really revokes its key.
//
// The memories follow the lifecycle suite's rules: the run marker, here the
// user's external id, lives only in the query, and the insight is plain prose
// about this SDK, distinct from every other write, so none reads as the same
// knowledge as another. A write a failing test leaves behind goes with its
// customer network.

import (
	"context"
	"crypto/tls"
	"errors"
	"fmt"
	"log/slog"
	"net"
	"os"
	"slices"
	"strings"
	"sync"
	"testing"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/mem"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/proto"

	adminv1 "github.com/memcoai/memcoai/go/internal/client/memcoai/admin/v1"
	authv1 "github.com/memcoai/memcoai/go/internal/client/memcoai/auth/v1"
	"github.com/memcoai/memcoai/go/internal/config"
	"github.com/memcoai/memcoai/go/memcoai"
)

var (
	opening = note{
		query: "How does the Memco Go SDK open a memory session as one of an organisation's own users?",
		title: "Go SDK impersonated sessions",
		content: "The Memco Go SDK opens a memory session as one of an organisation's own users when " +
			"StartSession is given the ExternalID option. It mints an impersonation key under the API " +
			"client's token, lists the domains under that key so the session learns its limits as that " +
			"user, and only then starts the session.\n\n" +
			"Every call made through the session, its tools and the memories it returns included, " +
			"carries that key rather than the client's token. Closing the session ends the key on the " +
			"service instead of leaving it live until it expires.",
	}
	renewal = note{
		query: "When does the Memco Go SDK renew an impersonation key?",
		title: "Go SDK impersonation key renewal",
		content: "The Memco Go SDK renews an impersonation key once four fifths of its lifetime have " +
			"passed. Every goroutine that finds the key due waits on one shared mint, so a burst of calls " +
			"mints a single key and a failed mint reaches all of them at once. The key it replaces is " +
			"ended only after the last call still carrying it has returned, so a renewal never revokes a " +
			"key under a call in flight.",
	}
	sweep = note{
		query: "What happens to open impersonation keys when a Memco Go SDK client closes?",
		title: "Go SDK impersonation key cleanup",
		content: "Closing a Memco Go SDK client waits for the calls in flight, then ends every " +
			"impersonation key a session left open under the client's own token, bounded by the client's " +
			"timeout rather than the caller's context, so a cancelled close never strands a key. A key the " +
			"service reports as already gone counts as ended, and one it cannot end is logged by its id, " +
			"never its value, and expires on its own.",
	}
)

// asked is the note with the run marker added to the query alone.
func asked(n note, marker string) note {
	n.query += " (system test " + marker + ")"
	return n
}

// carries reports whether a memory was written with marker in its query,
// which becomes one of its intents.
func carries(memory memcoai.Memory, marker string) bool {
	return slices.ContainsFunc(memory.Intents, func(intent string) bool { return strings.Contains(intent, marker) })
}

// frame is a message the proxy passes on unread.
type frame struct{ data []byte }

// raw carries frames as they are, under the proto codec's name, so neither
// end sees anything but protobuf.
type raw struct{}

func (raw) Marshal(v any) (mem.BufferSlice, error) {
	return mem.BufferSlice{mem.SliceBuffer(v.(*frame).data)}, nil
}

func (raw) Unmarshal(data mem.BufferSlice, v any) error {
	v.(*frame).data = data.Materialize()
	return nil
}

func (raw) Name() string { return "proto" }

// keyProxy stands between a client and the service, passing every call on as
// it is and keeping each impersonation key the service mints. The SDK shows a
// key's value nowhere, which is the point; only the value lets the test ask
// the service whether a closed session's key still works.
type keyProxy struct {
	address  string
	upstream *grpc.ClientConn

	mu   sync.Mutex
	keys []*adminv1.ImpersonationKey
}

func startProxy(t *testing.T) *keyProxy {
	t.Helper()
	cfg, err := config.Resolve(config.Input{Token: "unused"}, os.Getenv, slog.New(slog.DiscardHandler))
	if err != nil {
		t.Fatal(err)
	}
	transport := insecure.NewCredentials()
	if cfg.TLS {
		transport = credentials.NewTLS(&tls.Config{MinVersion: tls.VersionTLS12})
	}
	upstream, err := grpc.NewClient(cfg.DialTarget(), grpc.WithTransportCredentials(transport))
	if err != nil {
		t.Fatal(err)
	}
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	p := &keyProxy{address: listener.Addr().String(), upstream: upstream}
	server := grpc.NewServer(grpc.ForceServerCodecV2(raw{}), grpc.UnknownServiceHandler(p.forward))
	go func() { _ = server.Serve(listener) }()
	t.Cleanup(func() {
		server.Stop()
		_ = upstream.Close()
	})
	return p
}

func (p *keyProxy) forward(_ any, stream grpc.ServerStream) error {
	method, _ := grpc.MethodFromServerStream(stream)
	var request, response frame
	if err := stream.RecvMsg(&request); err != nil {
		return err
	}
	incoming, _ := metadata.FromIncomingContext(stream.Context())
	outgoing := metadata.MD{}
	if bearer := incoming.Get("authorization"); len(bearer) > 0 {
		outgoing.Set("authorization", bearer...)
	}
	ctx := metadata.NewOutgoingContext(stream.Context(), outgoing)
	if err := p.upstream.Invoke(ctx, method, &request, &response, grpc.ForceCodecV2(raw{})); err != nil {
		return err
	}
	if method == adminv1.AdminService_ImpersonateExternalUser_FullMethodName {
		key := &adminv1.ImpersonationKey{}
		if err := proto.Unmarshal(response.data, key); err != nil {
			return err
		}
		p.mu.Lock()
		p.keys = append(p.keys, key)
		p.mu.Unlock()
	}
	return stream.SendMsg(&response)
}

func (p *keyProxy) minted() []*adminv1.ImpersonationKey {
	p.mu.Lock()
	defer p.mu.Unlock()
	return slices.Clone(p.keys)
}

// through connects an API client to the service by way of the proxy.
func (p *keyProxy) through(t *testing.T) *memcoai.Client {
	t.Helper()
	administrator(t)   // skips without the API client's credentials
	plaintext := false // to the proxy on loopback; it dials the service as configured
	return connect(t, memcoai.Options{
		ClientID: os.Getenv(clientIDEnv), ClientSecret: os.Getenv(clientSecretEnv),
		Host: p.address, TLS: &plaintext,
	})
}

// endAgain ends an impersonation key directly, as the SDK would.
func (p *keyProxy) endAgain(ctx context.Context, externalID, keyID string) error {
	token, err := authv1.NewTokenServiceClient(p.upstream).IssueToken(ctx, &authv1.IssueTokenRequest{
		GrantType: "client_credentials", ClientId: os.Getenv(clientIDEnv), ClientSecret: os.Getenv(clientSecretEnv),
	})
	if err != nil {
		return err
	}
	authorized := metadata.AppendToOutgoingContext(ctx, "authorization", "Bearer "+token.GetAccessToken())
	_, err = adminv1.NewAdminServiceClient(p.upstream).EndImpersonation(authorized,
		&adminv1.EndImpersonationRequest{ExternalId: externalID, KeyId: keyID})
	return err
}

// keyWorks reports whether the service still takes key, by connecting with it.
func keyWorks(ctx context.Context, key *adminv1.ImpersonationKey) (bool, error) {
	client, err := memcoai.NewClient(memcoai.Options{Token: key.GetValue()})
	if err != nil {
		return false, err
	}
	defer func() { _ = client.Close(ctx) }()
	err = client.Connect(ctx)
	var rejected *memcoai.AuthenticationError
	if errors.As(err, &rejected) {
		return false, nil
	}
	return err == nil, err
}

// act writes, finds and reverts a memory as one user, failing if a search ever
// returns what other wrote. It returns rather than fails, so it can run on a
// goroutine of its own.
func act(ctx context.Context, client *memcoai.Client, domain, user, other string, fields note) error {
	session, err := client.Memory.StartSession(ctx, domain, memcoai.ExternalID(user))
	if err != nil {
		return fmt.Errorf("%s: opening: %w", user, err)
	}
	defer session.Close(ctx)
	before, err := session.Search(ctx, fields.query, memcoai.ScopedSearchParams{})
	if err != nil {
		return fmt.Errorf("%s: searching first: %w", user, err)
	}
	if slices.ContainsFunc(before.Memories, func(m memcoai.Memory) bool { return carries(m, user) }) {
		return fmt.Errorf("%s: its memory was found before it was written", user)
	}
	write, err := session.CreateMemory(ctx, memcoai.ScopedCreateMemoryParams{
		Query: fields.query, Title: fields.title, Content: fields.content, Source: memcoai.DataSourceAgent,
	})
	if err != nil {
		return fmt.Errorf("%s: writing: %w", user, err)
	}
	if write.OperationID == "" {
		return fmt.Errorf("%s: the write was accepted without an operation id to revert", user)
	}
	deadline := time.Now().Add(ingestTimeout)
	for pause := firstPoll; ; pause = min(2*pause, pollInterval) {
		result, err := session.Search(ctx, fields.query, memcoai.ScopedSearchParams{})
		if err != nil {
			return fmt.Errorf("%s: searching: %w", user, err)
		}
		if other != "" && slices.ContainsFunc(result.Memories, func(m memcoai.Memory) bool { return carries(m, other) }) {
			return fmt.Errorf("%s's session found what %s wrote, in a network it is not in", user, other)
		}
		if slices.ContainsFunc(result.Memories, func(m memcoai.Memory) bool { return carries(m, user) }) {
			break
		}
		if time.Now().Add(pause).After(deadline) {
			return fmt.Errorf("%s's memory never became searchable within %s", user, ingestTimeout)
		}
		time.Sleep(pause)
	}
	undo, err := session.RevertMemory(ctx, write.OperationID)
	if err != nil {
		return fmt.Errorf("%s: reverting: %w", user, err)
	}
	if undo.Outcome != memcoai.RevertOutcomeMemoryRemoved {
		return fmt.Errorf("reverting as %s, who wrote it, reported %s", user, undo.Outcome)
	}
	return nil
}

func TestASessionActsAsAnExternalUserAndItsKeyEndsWithIt(t *testing.T) {
	admin := administrator(t)
	network, user := placed(t, admin, rootNetwork(t, admin))
	proxy := startProxy(t)
	client := proxy.through(t)
	ctx := t.Context()
	t.Logf("[%s] acting as %s", network.Domain, user.ExternalID)

	session, err := client.Memory.StartSession(ctx, network.Domain, memcoai.ExternalID(user.ExternalID))
	if err != nil {
		t.Fatal(err)
	}
	if err := act(ctx, client, network.Domain, user.ExternalID, "", asked(opening, user.ExternalID)); err != nil {
		t.Fatal(err)
	}
	keys := proxy.minted()
	if len(keys) != 2 {
		t.Fatalf("two sessions minted %d keys", len(keys))
	}
	// The session act opened has closed; this one is still open, so its key
	// is refused below because of the close, not because it never worked.
	if works, err := keyWorks(ctx, keys[0]); err != nil || !works {
		t.Fatalf("the open session's key does not work: %v", err)
	}
	if works, err := keyWorks(ctx, keys[1]); err != nil || works {
		t.Fatalf("a closed session's key still works: %v", err)
	}
	session.Close(ctx)
	if works, err := keyWorks(ctx, keys[0]); err != nil || works {
		t.Fatalf("the key still works after its session closed: %v", err)
	}
	// Ending an ended key is answered NOT_FOUND, which the SDK counts as
	// ended: any other answer would have every close after a key ran out log
	// a warning, and stop the client's sweep at that key.
	if err := proxy.endAgain(ctx, user.ExternalID, keys[0].GetKeyId()); status.Code(err) != codes.NotFound {
		t.Fatalf("ending an ended key answered %v", err)
	}
	if len(proxy.minted()) != 2 {
		t.Fatalf("the sessions minted %d keys", len(proxy.minted()))
	}
}

func TestTwoUsersActAtOnceEachUnderAKeyOfItsOwn(t *testing.T) {
	admin := administrator(t)
	root := rootNetwork(t, admin)
	_, first := placed(t, admin, root)
	_, second := placed(t, admin, root)
	proxy := startProxy(t)
	client := proxy.through(t)
	ctx := t.Context()
	t.Logf("[%s] acting as %s and %s at once", root.Domain, first.ExternalID, second.ExternalID)

	errs := make(chan error, 2)
	go func() {
		errs <- act(ctx, client, root.Domain, first.ExternalID, second.ExternalID, asked(renewal, first.ExternalID))
	}()
	go func() {
		errs <- act(ctx, client, root.Domain, second.ExternalID, first.ExternalID, asked(sweep, second.ExternalID))
	}()
	for range 2 {
		if err := <-errs; err != nil {
			t.Error(err)
		}
	}
	// One key per session, each ended by its own session's close before the
	// client closes.
	keys := proxy.minted()
	if len(keys) != 2 {
		t.Fatalf("two sessions minted %d keys", len(keys))
	}
	for _, key := range keys {
		if works, err := keyWorks(ctx, key); err != nil || works {
			t.Errorf("%s still works after its session closed: %v", key.GetKeyId(), err)
		}
	}
}
